text/0000-project-descriptor-images.md

```diff
@@ -0,0 +1,159 @@
+# Meta
+[meta]: #meta
+- Name: Support specifying images in project descriptor
+- Start Date: 2021-10-26
+- Author(s): [@phil9909](https://github.com/phil9909), [@modulo11](https://github.com/modulo11), [@pbusko](https://github.com/pbusko), [@jkutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Add a new array of tables `[[images]]` to the `project.toml` file.
+
+# Definitions
+[definitions]: #definitions
+
+* **project descriptor**: The `project.toml` file that describes a project.
+* **image**: the output of a running a buildpack(s) against a project
+* **image name**: fully qualified name of an OCI image in the form `<registry>/<name>:<tag>`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, there is no support for specifying an output image in the `project.toml` although flags like `--tag` are already present.
+
+# What it is
```

Comment:
.@sclevine said that even though this is a lifecycle proposal, it would be good to see the `pack` experience. This is somewhat alluded to with the mention of `--tag` option.

Reply:
I believe the concerns come about the interoperability between config and provided inputs. How does any platform work with these values?

Take for example:

A `project.toml` has:

```toml
[[images]]
name = "spring-petclinic"
registry = "gcr.io"
tags = ["latest", "v1"]

[[images]]
name = "spring-petclinic"
tags = ["latest", "v1"]
```

and the user runs:

```bash
pack build my-app --tags my-app:v2
```

What is the expected outcome?

### Option 1

Order of precedence, any user-provided options override configuration. 

Following images are created:

1. `my-app` => `index.docker.io/my-app:latest`
1. `my-app:v2`  => `index.docker.io/my-app:v2`

### Option 2

Order of precedence, any configuration overrides user-provided options. 

Following images are created:

1. `gcr.io/spring-petclinic:latest`
1. `gcr.io/spring-petclinic:v1`
1. `spring-petclinic:latest`  => `index.docker.io/spring-petclinic:latest`
1. `spring-petclinic:v1`  => `index.docker.io/spring-petclinic:latest`

### Option 3

A union of user-provided and config.

Following images are created:

1. `my-app` => `index.docker.io/my-app:latest`
1. `my-app:v2`  => `index.docker.io/my-app:v2`
1. `gcr.io/spring-petclinic:latest`
1. `gcr.io/spring-petclinic:v1`
1. `spring-petclinic:latest`  => `index.docker.io/spring-petclinic:latest`
1. `spring-petclinic:v1`  => `index.docker.io/spring-petclinic:latest`

---

My vote would be Option 1 because it follows our current order of precedence strategy.

Something to consider is that arg1 (`<name>`) in `pack build` would now need to be optional to allow for `images[0]` value to take affect.

Reply:
I would vote for the Option 1.
Option 2 is hard to defend, if the user-provided explicit arguments to the `pack` are completely ignored.
Option 3 won't follow the currently established precedence strategy and will potentially provide unwanted results.

Reply:
I vote Option (4): build fails.

I think `pack build my-app` should mean "build my-app from the images table" when the images table exists.

Reply:
@jromero for Option 1, is `latest` image created because it's always created?

Reply:
> @jromero for Option 1, is `latest` image created because it's always created?

`latest` was created because the user provided `my-app` (without a tag) as a positional argument. The default when no tag is provided is `latest`.

Reply:
> I vote Option (4): build fails.
> 
> I think `pack build my-app` should mean "build my-app from the images table" when the images table exists.

I don't think this is ideal. @loewenstein gave a [good example of a use case](https://github.com/buildpacks/rfcs/pull/187#issuecomment-1006340692): 

> One use case for specifying the image in the pack call is tagging with the release version in CI. I don't think that we should or want to force users into modifying the project.toml in their release jobs.